File: java/Chapter 11/Question11_2/QuestionB.java
package Question11_2;

import java.util.Arrays;
import java.util.Hashtable;
import java.util.LinkedList;

import CtCILibrary.AssortedMethods;

public class QuestionB {
	public static String sortChars(String s) {
		char[] content = s.toCharArray();
		Arrays.sort(content);
		return new String(content);
	}
	
	public static void sort(String[] array) {
		Hashtable<String, LinkedList<String>> hash = new Hashtable<String, LinkedList<String>>();
		
		/* Group words by anagram */
		for (String s : array) {
			String key = sortChars(s); 
			if (!hash.containsKey(key)) {
				hash.put(key, new LinkedList<String>());
			}
			LinkedList<String> anagrams = hash.get(key);
			anagrams.push(s);
		}
		
		/* Convert hash table to array */
		int index = 0;
		for (String key : hash.keySet()) {
			LinkedList<String> list = hash.get(key);
			for (String t : list) {
				array[index] = t;
				index++;
			}
		}
	}
	
	public static void main(String[] args) {
		String[] array = {"apple", "banana", "carrot", "ele", "duck", "papel", "tarroc", "cudk", "eel", "lee"};
		sort(array);
		System.out.println(AssortedMethods.stringArrayToString(array));
	}
}

File: dart/Chapter11/Question11_2/questionB.dart

String sortChars(String s) {
  List<String> content = s.split('');
  content.sort();
  return content.join();
}

void sort(List<String> array) {
  Map<String, List<String>> hash = new Map<String, List<String>>();

  /* Group words by anagram */
  for (String s in array) {
    String key = sortChars(s);
    if (!hash.containsKey(key)) {
      hash[key] = new List<String>();
    }
    List<String> anagrams = hash[key];
    anagrams.add(s);
  }

  /* Convert hash table to array */
  int index = 0;
  for (String key in hash.keys) {
    List<String> list = hash[key];
    for (String t in list) {
      array[index] = t;
      index++;
    }
  }
}

void main() {
  List<String> array = ["apple", "banana", "carrot", "ele", "duck", "papel", "tarroc", "cudk", "eel", "lee"];
  print(array);
  sort(array);
  print(array);
}